src/Collection/Collection.service.ts

```diff
@@ -23,18 +32,210 @@ import {
+    }
+  }
+
+  /**
+   * It should receive the list of items to be operated on so it can decide where it should create or update the curations for them accordingly
+   * It will also receive a signed message and its signature, so it can check and store it in the SlotUsageCheque
+   */
+  public async publishTPCollection(
+    dbCollection: ThirdPartyCollectionAttributes,
+    dbItems: ItemAttributes[],
+    signedMessage: string,
+    signature: string
+  ): Promise<{
+    collection: CollectionAttributes
+    items: FullItem[]
+  }> {
+    // For DCL collections, once a published collection item changes, the PUSH CHANGES button appears
+    // That will fire a /collections/${collectionId}/curation which will create a new CollectionCuration
+    // Subsequent changes will not show the push changes button, as it's already under_review
+
+    // For TP items, creations always exist. PUSH CHANGES should appear if the item has an approved ItemCuration and has changes in the Catalyst
+    // That should fire /items/:id/curation for each item that changed
+
+    // What we should migrate over here is the creation of the virtual CollectionCuration, as there'll always be a publish before a PUSH CHANGES
+
+    if (dbItems.length === 0) {
+      throw new InvalidRequestError('Tried to publish no TP items')
+    }
+
+    try {
+      ethers.utils.verifyMessage(signedMessage, signature) // Throws if invalid
+    } catch (error) {
+      throw new InvalidRequestError(
+        'Tried to publish TP items with an invalid signed message or signature'
+      )
+    }
+
+    const collectionId = dbItems[0].collection_id
+    if (!collectionId) {
+      throw new InvalidRequestError(
+        'Cannot publish items without a collection id'
+      )
+    }
+
+    const allTheSameCollection = dbItems.every(
+      (item) => item.collection_id === collectionId
+    )
+    if (!allTheSameCollection) {
+      throw new InvalidRequestError(
+        'Cannot publish items that belong to different collections'
+      )
+    }
+
+    const isPublished = await ItemCuration.findLastCreatedByCollectionIdAndStatus(
+      collectionId,
+      CurationStatus.PENDING
+    )
+    if (isPublished) {
+      throw new AlreadyPublishedCollectionError(
+        collectionId,
+        CollectionType.THIRD_PARTY,
+        CollectionAction.UPDATE
+      )
+    }
+
+    const now = new Date()
+    await SlotUsageCheque.create<SlotUsageChequeAttributes>({
+      id: uuid(),
+      signedMessage,
+      collection_id: dbCollection.id,
+      third_party_id: dbCollection.third_party_id,
+      created_at: now,
+      updated_at: now,
+    })
+
+    const promises = []
+    for (const item of dbItems) {
+      promises.push(
+        ItemCuration.create<ItemCurationAttributes>({
+          id: uuid(),
+          item_id: item.id,
+          status: CurationStatus.PENDING,
+          created_at: now,
+          updated_at: now,
+        })
+      )
+    }
```

Comment:
This is a suggestion, but I think that the `map` is perfect for this job as it eliminates the need of doing pushes.
```suggestion
    const promises = dbItems.map(item => 
        ItemCuration.create<ItemCurationAttributes>({
          id: uuid(),
          item_id: item.id,
          status: CurationStatus.PENDING,
          created_at: now,
          updated_at: now,
        })
    )
```

Reply:
To create the curations, we could move the `insertCuration` method that we've got in the `Curation.router` into a `Curation.service` and use that.

Reply:
Heh, [there was a comment above this line](https://github.com/decentraland/builder-server/pull/409/commits/9fffc33ff4d5f777ea306d725526ba8a642f9673#diff-0d2b44325541c42451d1866d04950fb50ffcc55ba4c50ec4db19ad16bdd7ef44L206) saying that exact same thing. I figured that it wasn't really necessary as `create` is typed and `insertCuration` had code like this:

```ts
    const curationService = CurationService.byType(type)

    await this.validateAccessToCuration(curationService, ethAddress, id)

    const curation = await curationService.getLatestById(id)
```

which is not really relevant here I think

Reply:
The access check is not relevant, but checking if there's another curation that's not pending might be (at least in other situation). Saying so, we can see if it needs a refactor later.